lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -442,10 +441,17 @@ swift::matchWitness(
   if (req->getKind() != witness->getKind())
     return RequirementMatch(witness, MatchKind::KindConflict);
 
+  // If the witness has not been validated yet, do so now.
+  if (!witness->hasValidSignature()) {
+    auto &ctx = dc->getASTContext();
+    ctx.getLazyResolver()->resolveDeclSignature(witness);
+  }
+
   // If the witness is invalid, record that and stop now.
   if (witness->isInvalid())
     return RequirementMatch(witness, MatchKind::WitnessInvalid);
 
+  // If we're currently validating the witness, bail out.
```

Comment:
This comment seems like it shouldn't be correct. Either this function is called re-entrantly and will loop infinitely because of the above call, or the above call notices the circularity and returns without setting anything up. If you're talking about the latter, maybe this should move inside the first `if` to make it clearer what's going on?

Reply:
Right, it's not this function that's called re-entrantly; the bad case is when validateDecl() calls us, and we call validateDecl(). In this case validateDecl() returns without doing anything.